### PR TITLE
Fix instrumentation of SQLAlchemy when using sqlalchemy.engine_from_config

### PR DESCRIPTION
**Pull request by moredip**
# Description

Fixes a bug where SQLAlchemy engines created via `sqlalchemy.engine_from_config(...)` are not instrumented by `SQLAlchemyInstrumentor`.

`sqlalchemy.engine_from_config(...)` directly calls `create_engine(...)` imported from `sqlalchemy.engine.create`. If we don't wrap that copy of the `create_engine` method then engines created via `engine_from_config` are not
instrumented.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- Added a unit test
-  ran it and checked that it failed
- made the fix
- verified that the test now passes

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -40,6 +40,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   ([#2792](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2792))
 - `opentelemetry-instrumentation-tornado` Handle http client exception and record exception info into span
   ([#2563](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2563))
+- `opentelemetry-instrumentation-sqlalchemy` Fixes engines from sqlalchemy.engine_from_config(...) not being fully instrumented
```
Comment (emdneto) [inline CHANGELOG.md:43]:
```suggestion
- `opentelemetry-instrumentation-sqlalchemy` Fixes engines from `sqlalchemy.engine_from_config` not being fully instrumented
```

**Review by xrmx — APPROVED** [review]

**Review by emdneto — APPROVED** [review]

**Review by emdneto — CHANGES_REQUESTED** [review]
oh wait, there's a pypy test failing 😓

**Comment by tammy-baylis-swi** [comment]
Thanks for putting this in! Currently 5 of the unit tests are failing with `ModuleNotFoundError: No module named 'sqlalchemy.engine.create'` which comes from `sqlalchemy.engine.create` being introduced in [SQLAlchemy 2](https://docs.sqlalchemy.org/en/20/core/engines.html) but the tests are run in [1.1](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/6a54106f5b39ac4d09482498603c1246f4097c2d/instrumentation/opentelemetry-instrumentation-sqlalchemy/test-requirements-0.txt#L10) and [1.4](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/6a54106f5b39ac4d09482498603c1246f4097c2d/instrumentation/opentelemetry-instrumentation-sqlalchemy/test-requirements-0.txt#L10) (also [tox.ini comment](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/6a54106f5b39ac4d09482498603c1246f4097c2d/tox.ini#L259-L260)). The instrumentor doesn't seem to have bounds for [supported SQLAlchemy version](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/6a54106f5b39ac4d09482498603c1246f4097c2d/instrumentation/opentelemetry-instrumentation-sqlalchemy/pyproject.toml#L36-L38) though, so I think it would be ok to bump the `test-requirements` versions. Please could you give that a try.

**Comment by moredip** [comment]
> Thanks for putting this in! Currently 5 of the unit tests are failing with `ModuleNotFoundError: No module named 'sqlalchemy.engine.create'` which comes from `sqlalchemy.engine.create` being introduced in [SQLAlchemy 2](https://docs.sqlalchemy.org/en/20/core/engines.html) but the tests are run in [1.1](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/6a54106f5b39ac4d09482498603c1246f4097c2d/instrumentation/opentelemetry-instrumentation-sqlalchemy/test-requirements-0.txt#L10) and [1.4](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/6a54106f5b39ac4d09482498603c1246f4097c2d/instrumentation/opentelemetry-instrumentation-sqlalchemy/test-requirements-0.txt#L10) (also [tox.ini comment](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/6a54106f5b39ac4d09482498603c1246f4097c2d/tox.ini#L259-L260)). The instrumentor doesn't seem to have bounds for [supported SQLAlchemy version](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/6a54106f5b39ac4d09482498603c1246f4097c2d/instrumentation/opentelemetry-instrumentation-sqlalchemy/pyproject.toml#L36-L38) though, so I think it would be ok to bump the `test-requirements` versions. Please could you give that a try.

Sorry for dropping the ball on this for a while, @tammy-baylis-swi. I've not contributed to otel before, and not used tox before either; would you be able to elaborate on what I need to do with the `test-requirements` versions?

**Comment by tammy-baylis-swi** [comment]
> Sorry for dropping the ball on this for a while, @tammy-baylis-swi. I've not contributed to otel before, and not used tox before either; would you be able to elaborate on what I need to do with the `test-requirements` versions?

No worries @moredip ! Coincidentally, yesterday we merged 2 PRs to address some of what I mentioned:

* https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2971
* https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2976

I'm going to merge `main` into this branch and reassess.

File: instrumentation/opentelemetry-instrumentation-sqlalchemy/tests/test_sqlalchemy.py [hunk 2]
```diff
@@ -182,6 +182,17 @@ def test_create_engine_wrapper(self):
             "opentelemetry.instrumentation.sqlalchemy",
         )
 
+    def test_instrument_engine_from_config(self):
+        SQLAlchemyInstrumentor().instrument()
+        from sqlalchemy import engine_from_config  # pylint: disable-all
+
+        engine = engine_from_config({"sqlalchemy.url": "sqlite:///:memory:"})
+        cnx = engine.connect()
+        cnx.execute("SELECT	1 + 1;").fetchall()
```
Comment (tammy-baylis-swi) [inline instrumentation/opentelemetry-instrumentation-sqlalchemy/tests/test_sqlalchemy.py:191]: This should fix some current `sqlalchemy.exc.ObjectNotExecutableError` in the tests:

```suggestion
        cnx.execute(text("SELECT	1 + 1;")).fetchall()
```

File: instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/__init__.py [hunk 3]
```diff
@@ -181,6 +181,13 @@ def _instrument(self, **kwargs):
                 tracer, connections_usage, enable_commenter, commenter_options
             ),
         )
+        _w(
+            "sqlalchemy.engine.create",
+            "create_engine",
+            _wrap_create_engine(
+                tracer, connections_usage, enable_commenter, commenter_options
+            ),
+        )
```
Comment (tammy-baylis-swi) [inline instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/__init__.py:190]: We're getting `ModuleNotFoundError: No module named 'sqlalchemy.engine.create'` from the [unit tests](https://github.com/open-telemetry/opentelemetry-python-contrib/actions/runs/11751268984/job/32742678868?pr=2816) and that comes from this instrumentor [supporting SQLAlchemy 1.1](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-sqlalchemy/test-requirements-0.txt#L9), which does not have the `sqlalchemy.engine.create` module.

Please could you try something like what's below on L196 and see if that helps (not sure if it was introduced in 1.4 or 2.0):

```suggestion
        if parse_version(sqlalchemy.__version__).release >= (1, 4):
            _w(
                "sqlalchemy.engine.create",
                "create_engine",
                _wrap_create_engine(
                    tracer, connections_usage, enable_commenter, commenter_options
                ),
            )
```

Comment (emdneto) [reply instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/__init__.py:190]: I would also leave a comment

File: instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/__init__.py [hunk 4]
```diff
@@ -224,6 +231,7 @@ def _instrument(self, **kwargs):
     def _uninstrument(self, **kwargs):
         unwrap(sqlalchemy, "create_engine")
         unwrap(sqlalchemy.engine, "create_engine")
+        unwrap(sqlalchemy.engine.create, "create_engine")
```
Comment (emdneto) [inline instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/__init__.py:234]: Should we move this to the if statement `parse_version(sqlalchemy.__version__).release >= (1, 4)` as well?

Comment (tammy-baylis-swi) [reply instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/__init__.py:234]: Agreed!

**Comment by tammy-baylis-swi** [comment]
Getting closer! Please could you run a `tox -e ruff` on your local and push the formatting change to get this to pass:

https://github.com/open-telemetry/opentelemetry-python-contrib/actions/runs/11960954700/job/33346209372?pr=2816

File: CHANGELOG.md [hunk 5]
```diff
@@ -129,6 +129,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   ([#2792](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2792))
 - `opentelemetry-instrumentation-tornado` Handle http client exception and record exception info into span
   ([#2563](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2563))
+- `opentelemetry-instrumentation-sqlalchemy` Fixes engines from `sqlalchemy.engine_from_config` not being fully instrumented
```
[the remark this answers is not in the document]
Comment (emdneto) [reply CHANGELOG.md:132]: Move to # Unreleased section

File: CHANGELOG.md [hunk 6]
```diff
@@ -38,6 +38,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   ([#3022](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/3022))
 - Replace all instrumentor unit test `assertEqualSpanInstrumentationInfo` calls with `assertEqualSpanInstrumentationScope` calls
   ([#3037](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/3037))
+- `opentelemetry-instrumentation-sqlalchemy` Fixes engines from `sqlalchemy.engine_from_config`not being fully instrumented
```
Comment (emdneto) [inline CHANGELOG.md:41]:
```suggestion
- `opentelemetry-instrumentation-sqlalchemy` Fixes engines from `sqlalchemy.engine_from_config` not being fully instrumented
```

**Review by emdneto — APPROVED** [review]

**Review by tammy-baylis-swi — APPROVED** [review]
Looks good to me. Thank you 😺

**Comment by moredip** [comment]
> Looks good to me. Thank you 😺

Thanks so much for patiently shepherding me through! Very much appreciated 🙇

**Comment by moredip** [comment]
@tammy-baylis-swi @emdneto @xrmx is this OK to merge, or do I also need to wait on a 👍 from @shalevr

**Review by shalevr — APPROVED** [review]

**Pull request merged**